src/components/Template/Template.tsx

```diff
@@ -1,23 +1,65 @@
 /** @jsx h */
 
 import { h, Component } from 'preact';
-import PropTypes from 'prop-types';
 import { renderTemplate, isEqual } from '../../lib/utils';
 
-class Template extends Component {
-  shouldComponentUpdate(nextProps) {
+type TemplateProps = {
+  data?: Record<string, any>;
+  rootProps?: Record<string, any>;
+  rootTagName?: string;
+  templateKey?: string;
+  templates?: Record<
+    string,
+    | string
+    | ((data: any, bindEvent: (...args: any[]) => string) => string)
+    | undefined
+  >;
+  templatesConfig?: {
+    helpers?: Record<
+      string,
+      (text: string, render: (value: any) => string) => string
+    >;
+    // https://github.com/twitter/hogan.js/#compilation-options
+    compileOptions?: {
+      asString?: boolean;
+      sectionTags?: Array<{
+        o?: string;
+        c?: string;
+      }>;
+      delimiters?: string;
+      disableLambda?: boolean;
+    };
+  };
+  useCustomCompileOptions?: Record<string, boolean | undefined>;
+  bindEvent?: (...args: any[]) => string;
+} & typeof defaultProps;
+
+const defaultProps = Object.freeze({
+  data: {},
+  rootTagName: 'div',
+  useCustomCompileOptions: {},
+  templates: {},
+  templatesConfig: {},
+});
+
+class Template extends Component<TemplateProps> {
+  public static readonly defaultProps = defaultProps;
+
+  public shouldComponentUpdate(nextProps: TemplateProps) {
     return (
       !isEqual(this.props.data, nextProps.data) ||
       this.props.templateKey !== nextProps.templateKey ||
       !isEqual(this.props.rootProps, nextProps.rootProps)
     );
   }
 
-  render() {
-    const RootTagName = this.props.rootTagName;
-    const useCustomCompileOptions = this.props.useCustomCompileOptions[
-      this.props.templateKey
-    ];
+  public render() {
+    const RootTagName = (this.props.rootTagName as unknown) as (
+      props: h.JSX.HTMLAttributes
+    ) => h.JSX.Element;
```

Comment:
isn't it a string though? I'm pretty sure you should be allowed to start a string jsx element, no?

Reply:
All string are not allowed (it must match an HTML Element tag name), but I found a better solution to avoid casting:

```ts
rootTagName?: keyof h.JSX.IntrinsicElements;
```